Libs/libMacGitverCore/Widgets/ProgressDlg.cpp

```diff
@@ -0,0 +1,206 @@
+
+#include "ProgressDlg.hpp"
+#include "ui_ProgressDlg.h"
+#include "ui_ProgressWdgt.h"
+
+#include <QStringBuilder>
+#include <QCloseEvent>
+#include <QPushButton>
+
+namespace Private
+{
+
+    class ProgressWdgt : public QWidget, public Ui::ProgressWdgt
+    {
+    public:
+        typedef QMap< QString, QPointer<ProgressWdgt> > Steps;
+
+        enum Status { Running = 0, Stopped };
+
+    public:
+        ProgressWdgt(const QString& description)
+        {
+            setupUi(this);
+            progressBar->setMinimum(0);
+            progressBar->setMaximum(100);
+            txtHeader->setText( description );
+        }
+
+    public:
+        Steps   mSteps;
+        Status  mStatus     = Running;
+
+        qreal   mPercentage = 0.;
+    };
+
+}
+
+
+ProgressDlg::ProgressDlg()
+    : BlueSky::Dialog()
+    , ui( new Ui::ProgressDlg )
+    , mDone( false )
+{
+    ui->setupUi( this );
+
+    QPushButton* close = ui->buttonBox->button( QDialogButtonBox::Close );
+    close->setEnabled( false );
+    connect( close, &QPushButton::clicked, this, &ProgressDlg::close );
+
+    QPalette p;
+    p.setColor( QPalette::Base, p.color( QPalette::Window ) );
+    p.setColor( QPalette::Text, p.color( QPalette::WindowText ) );
+    ui->txtLog->setPalette( p );
+
+    connect(&mUpdater, &QTimer::timeout, this, &ProgressDlg::updateActivities);
+    mUpdater.start(250);
+}
+
+ProgressDlg::~ProgressDlg()
+{
+    delete ui;
+}
+
+int ProgressDlg::updateInterval() const
+{
+    return mUpdater.interval();
+}
+
+void ProgressDlg::setUpdateInterval(int msec)
+{
+    mUpdater.setInterval(msec);
+}
+
+void ProgressDlg::addActivity(const QString& description, QObject* activity,
+                              const StepInfo::List& steps)
+{
+    Q_ASSERT(activity);
+
+    Private::ProgressWdgt* a( new Private::ProgressWdgt(description) );
```

Comment:
Using the Copy-CTor of a pointer seems wired at least. I'd prefer to see an assignment here (at least to show clearly what happens - You know that C++ has strange rules on calling move/copy/assignment ctor/operator regardless if the codes uses a ctor or assignment)...